### PR TITLE
Fix/Update TorchConnector tutorial

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
This PR fixes the gradient propagation in Part 2 of the TorchConnector tutorial, and updates the related docstrings for the `TorchConnector`, `OpflowQNN`, `CircuitQNN`, `TwoLayerQNN` and `SamplingQNN ` classes to avoid future mistakes. It also updates the tutorial's explanations to add extra context and include relevant warnings.

### Details and comments
- The cause of the problem was not setting ` input_gradients=True` at the NN constructor to allow for gradient propagation.
- An additional fix was required to adjust gradient dimensions (`lines 177-179` of `torch_connector.py`). 
- The current code's data loader doesn't work with the newest version of torchvision (0.10.0).
- The heading levels are hopefully fixed too (PR #139 )

### CODE REVIEW

**Comment by ElePT** [comment]
Just for reference, the reason why torchvision 0.10.0 does not work is related to the ToTensor transform and Pillow, as shown in https://github.com/pytorch/vision/issues/4146. It might get fixed with version 0.10.1. 

If we really want to use torchvision 0.10.0, we actually can, but we need to make sure that Pillow <= 8.2.0.

File: docs/tutorials/05_torch_connector.ipynb [hunk 1]
```diff
@@ -262,120 +329,142 @@
   },
   {
    "cell_type": "markdown",
-   "id": "moral-berry",
+   "id": "fb029de7",
    "metadata": {},
    "source": [
-    "The red circles indicate wrongly classified data points."
+    "We can see that the model can generally achieve an accuracy over 75% (depending on the dataset). The red circles indicate wrongly classified data points."
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:335]: Can we be sure we can achieve 75%?

File: docs/tutorials/05_torch_connector.ipynb [hunk 2]
```diff
@@ -912,38 +1116,20 @@
     "        count += 1"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "e6a26b82",
+   "metadata": {},
+   "source": [
+    " 🎊🎊 **You are now able to experiment with your own hybrid datasets and architectures using Qiskit Machine Learning. Good Luck!** 🎊🎊"
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:1124]: It is unclear to me what is exactly these emoji mean :)

**Review by adekusar-drl — COMMENTED** [review]
Looks good. 
Unclear why the changes are required in `TorchConnector` (indices). 
Also, please, clean up the notebook: run all cells to get proper results(e.g. if I just open the notebook and want to have a look at the output), remove empty cells.

**Comment by woodsp-ibm** [comment]
For the notebook can you commit a version that has run to completion. The one here it looks as though you stopped it towards the end as it has a keyboard interrupt exception printed in it. In CI does run to the end - the output though mis-predicts every one.
```
Performance on test data:
        Loss: -0.7654
        Accuracy: 0.0%
```
You can see the build output, where we generate the html. If you look at where the status is at the bottom of the page here and says **All checks have passed** there is `Show all checks` clickable on the right. Do that and click on any one of the `Details` clickables. It will go to another screen. Above the right panel, that shows the info for a job, you will see `Artifacts` drop down (shows numeric 8 as there are eight we create). These are build output like the html which is created from the notebooks etc. If you click on the dropdown arrow then download `tutorials3.8` at the end. Unpack the tar.gz file that downloads and you will see an `index.html`. This is the root page of the docs that we would publish and you will find the html that was created from the tutorial.

In that tutorial html you can see several aspects do not come out so well e.g.
```
ptimizer algorithms defined in the `torch.optim <https://pytorch.org/docs/stable/optim.html>`__ package.
```
the above link - maybe the backticks on torch.optim end up causing the issue for when its translated to sphinx.

Also several headings are not being converted e.g
```
#### A. Classification with PyTorch and 
```
and
```
### 2. Regression
```
for example.

File: qiskit_machine_learning/neural_networks/circuit_qnn.py [hunk 3]
```diff
@@ -87,9 +87,11 @@ def __init__(
                 is provided.
             gradient: The gradient converter to be used for the probability gradients.
             quantum_instance: The quantum instance to evaluate the circuits. Note that
-                if ``sampling==True``, 'statevector_simulator' is not a valid backend for the
+                if ``sampling==True``, a statevector simulator is not a valid backend for the
                 quantum instance.
             input_gradients: Determines whether to compute gradients with respect to input data.
+                Note that this parameter is ``False`` by default, and must be explicitly set to
+                ``True`` for a proper gradient computation when using ``TorchConnector``.
```
[the remark this answers is not in the document]
Comment (woodsp-ibm) [reply qiskit_machine_learning/neural_networks/circuit_qnn.py:94]: I have to wonder whether we could do better here for the end-user. If a user 'must' set it - can we set it on their behalf instead? Or at least check the value and fail in some way in the TorchConnector if it 'must be set'.

File: docs/tutorials/05_torch_connector.ipynb (hunk 2 above)
Comment (ElePT) [reply docs/tutorials/05_torch_connector.ipynb]: Hahaha I would say that they are the Japanese version of 🎉. I will change them for 🎉  just in case.

File: docs/tutorials/05_torch_connector.ipynb (hunk 1 above)
Comment (ElePT) [reply docs/tutorials/05_torch_connector.ipynb]: Actually, I did multiple runs this morning and no, we cannot be sure. I removed this line on the new commit.

File: qiskit_machine_learning/neural_networks/circuit_qnn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ElePT) [reply qiskit_machine_learning/neural_networks/circuit_qnn.py:94]: I didn't add them to the base class because (correct me if I'm wrong) I thought that the instructions wouldn't show up directly when looking up i.e. "CircuitQNN". And I found it convoluted to have to refer people to the base class for something that is so "practically used". 

Adding a check makes a lot of sense! (If not changing the way it's done, I am not 100% sure of the reasons behind this parameter being set like this, but I remember that Anton changed it multiple times, maybe we should look at it further).

[the remark this answers is not in the document]
Comment (adekusar-drl) [reply qiskit_machine_learning/neural_networks/circuit_qnn.py:94]: @woodsp-ibm when I introduced `input_gradients`, they were set to `True` in `TorchConnector` by default. But then we discussed this with @stefan-woerner and decided that gradients with respect to input are required only in hybrid networks and thus they should be turned on explicitly. Well, we can revert this behavior and make them enabled by default again.

File: qiskit_machine_learning/connectors/torch_connector.py [hunk 4]
```diff
@@ -172,7 +173,7 @@ def backward(ctx: Any, grad_output: Tensor) -> Tuple:  # type: ignore
                 # pass) j for each observation i in the batch. Multiplies this with the gradient
                 # from this point on backwards with respect to each input k. Sums over all i and
                 # j to get total gradient of output w.r.t. each input k.
-                input_grad = einsum("ij,ijk->k", grad_output, input_grad)
+                input_grad = einsum("ij,ijk->ik", grad_output, input_grad)
```
[the remark this answers is not in the document]
Comment (ElePT) [reply qiskit_machine_learning/connectors/torch_connector.py]: I changed the comment to try to explain this. Basically the input gradients depend on the batch dimension, but the weight gradients depend only on the layer parameter shape you determine at the beginning, that is why I added the index ` i` only for the `input_grad`. Without this, the dimensions don't match during  backprop for batch sizes larger than 1.

[the remark this answers is not in the document]
Comment (ElePT) [reply qiskit_machine_learning/connectors/torch_connector.py]: In case you were wondering, the unit tests give the same results because during the check you end up summing all the gradients anyway.

File: qiskit_machine_learning/neural_networks/circuit_qnn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (woodsp-ibm) [reply qiskit_machine_learning/neural_networks/circuit_qnn.py:94]: Ok, as I mentioned its more to help an end user get things right - like the warnings that were added recently around output shape behavior. Logging a warning if this was detected could be helpful - though perhaps its hard to detect its use in a hybrid case specifically. We can see how things go though as it is, whether we get feedback from users etc.

**Comment by woodsp-ibm** [comment]
Since there is a bug fix of the TorchConnector in with this I think we should have a reno for the release notes for this.

File: qiskit_machine_learning/connectors/torch_connector.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adekusar-drl) [reply qiskit_machine_learning/connectors/torch_connector.py]: I see. But when gradients were computed without `i` I did not see an error in backdrop.

**Comment by adekusar-drl** [comment]
@ElePT Can you please in the cell 8 replace:
```python
    loss = 0.0                                             # Initialize loss    
    for x, y_target in zip(X, y01):                        # Evaluate batch loss
        output = model2(Tensor(x)).reshape(1, 2)           # Forward pass
        loss += f_loss(output, Tensor([y_target]).long())  # Calculate loss
```
with:
```python
    loss = f_loss(model2(X_), y01_)
```
This is a batch loss computation instead of iterating over all elements.

File: qiskit_machine_learning/connectors/torch_connector.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ElePT) [reply qiskit_machine_learning/connectors/torch_connector.py]: Yes, I was also wondering how did the MNIST tutorial manage to run originally. My theory is that it's related to the new PyTorch release (1.9.0), that came out in June 15. 

Before, I think that torch was broadcasting the input_grad tensors to "compensate" for the missing batch dimension during backprop. And because the original tutorial only used batch size 1, everything worked fine. The new PyTorch release has changed some aspects of broadcasting, so it's most likely not possible to do this anymore (and I think that it would not have worked well for batch size >1 anyway).  I don't know if there is any test with hybrid networks and batch size >=2, but I haven't found it. Actually, now that I think about it, it would make a lot of sense to include a few hybrid nn unit tests...

**Comment by adekusar-drl** [comment]
Looks good to me, thanks a lot. One thing is left - the dataset files. I've already pinged @manoelmarques.

**Review by adekusar-drl — APPROVED** [review]
Thanks a lot!

**Review by woodsp-ibm — APPROVED** [review]
